language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1201 @@
+    fn push(&mut self, frame: Frame<'txn, FunctionRef<'txn>>) {
+        self.0.push(frame);
+    }
+
+    /// Pop a `Frame` off the call stack.
+    fn pop(&mut self) -> Option<Frame<'txn, FunctionRef<'txn>>> {
+        self.0.pop()
+    }
+}
+
+/// A `Frame` is the execution context for a function. It holds the locals of the function and
+/// the function itself.
+struct Frame<'txn, F: 'txn> {
+    pc: u16,
+    locals: Locals,
+    function: F,
+    phantom: PhantomData<&'txn F>,
+}
+
+/// An `ExitCode` from `execute_code_unit`.
+#[derive(Debug)]
+enum ExitCode {
+    /// A `Return` opcode was found.
+    Return,
+    /// A `Call` opcode was found.
+    Call(FunctionHandleIndex),
+    /// A `CreateAccount` opcode was found.
+    CreateAccount,
+}
+
+impl<'txn, F> Frame<'txn, F>
+where
+    F: FunctionReference<'txn>,
+{
+    /// Create a new `Frame` given a `FunctionReference` and the function `Locals`.
+    ///
+    /// The locals must be loaded before calling this.
+    fn new(function: F, locals: Locals) -> Self {
+        Frame {
+            pc: 0,
+            locals,
+            function,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Return the code stream of this function.
+    fn code_definition(&self) -> &'txn [Bytecode] {
+        self.function.code_definition()
+    }
+
+    /// Return the `LoadedModule` this function lives in.
+    fn module(&self) -> &'txn LoadedModule {
+        self.function.module()
+    }
+
+    /// Copy a local from this frame at the given index. Return an error if the index is
+    /// out of bounds or the local is `Invalid`.
+    fn copy_loc(&self, idx: LocalIndex) -> VMResult<Value> {
+        self.locals.copy_loc(idx as usize)
+    }
+
+    /// Move a local from this frame at the given index. Return an error if the index is
+    /// out of bounds or the local is `Invalid`.
+    fn move_loc(&mut self, idx: LocalIndex) -> VMResult<Value> {
+        self.locals.move_loc(idx as usize)
+    }
+
+    /// Store a `Value` into a local at the given index. Return an error if the index is
+    /// out of bounds.
+    fn store_loc(&mut self, idx: LocalIndex, value: Value) -> VMResult<()> {
+        self.locals.store_loc(idx as usize, value)
+    }
+
+    /// Borrow a local from this frame at the given index. Return an error if the index is
+    /// out of bounds or the local is `Invalid`.
+    fn borrow_loc(&mut self, idx: LocalIndex) -> VMResult<Value> {
+        self.locals.borrow_loc(idx as usize)
+    }
+}
+
+//
+// Below are all the functions needed for gas synthesis and gas cost.
+// The story is going to change given those functions expose internals of the Interpreter that
+// should never leak out.
+// For now they are grouped in a couple of temporary struct and impl that can be used
+// to determine what the needs of gas logic has to be.
+//
+
+pub struct InterpreterForGasCost<'a, 'alloc, 'txn>
```

Comment:
Can we guard this struct with feature flag so that we won't see it in normal compilation flow?

Reply:
I am assuming you mean next one given this is vital to the functioning of gas.
I did that though, as mentioned in the initial comment, I would rather leave it as is to convey how much that code has to be changed... ;)
hopefully nothing breaks...